### PR TITLE
Gdb pretty printers

### PR DESCRIPTION
**Pull request by krux02**
For the most part, it puts a nim-gdb in the bin directory that starts gdb with the correct argument to also load the python module so that debugging Nim on gdb is a charm. if you can say that debugging with gdb is a charm at all.

I marked the debug proc with an exportc progma to control how that function is seen for gdb. gdb allows to call C functions and writing 'call debugSym(s)' to gdb works much better than 'call debug_<isertlongHash>'. Other than that it should the exportc pragme should not have any side effects.

### CODE REVIEW

File: compiler/astalgo.nim [hunk 1]
```diff
@@ -27,9 +27,9 @@ proc lineInfoToStr*(conf: ConfigRef; info: TLineInfo): Rope
 when declared(echo):
   # these are for debugging only: They are not really deprecated, but I want
   # the warning so that release versions do not contain debugging statements:
-  proc debug*(n: PSym; conf: ConfigRef = nil) {.deprecated.}
-  proc debug*(n: PType; conf: ConfigRef = nil) {.deprecated.}
-  proc debug*(n: PNode; conf: ConfigRef = nil) {.deprecated.}
+  proc debug*(n: PSym;  conf: ConfigRef = nil) {.exportc: "debugSym", deprecated.}
```
Comment (Araq) [inline compiler/astalgo.nim]: Remove this space. We don't play Ascii-Art alignment games in the compiler.

File: tests/untestable/gdb/prettyPrinterTestInvoke [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
Comment (Araq) [inline tests/untestable/gdb/prettyPrinterTestInvoke]: The filename should be ``gdb_pretty_printer_test.sh``.

File: tests/untestable/gdb/prettyPrinterTest.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import gdb
```
Comment (Araq) [inline tests/untestable/gdb/prettyPrinterTest.py]: The filename should be ``gdb_pretty_printer_test.py``

File: bin/nim-gdb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env sh
```
Comment (Araq) [inline bin/nim-gdb]: This should be a Nim program.

File: tests/untestable/gdb/prettyPrinterTestProgram.nim [hunk 5]
```diff
@@ -0,0 +1,53 @@
+
```
Comment (Araq) [inline tests/untestable/gdb/prettyPrinterTestProgram.nim]: ``gdb_pretty_printer_test.nim``

Comment (krux02) [reply tests/untestable/gdb/prettyPrinterTestProgram.nim]: no problem

File: tests/untestable/gdb/prettyPrinterTestInvoke (hunk 2 above)
Comment (krux02) [reply tests/untestable/gdb/prettyPrinterTestInvoke]: Well I checked it, and a very common convention for executable shell scripts on Linux is to leave away the file extension. The file type is then specified with the ``#!`` line at the beginning of the file. Editors are pretty good at detecting the file format this way. Shell libraries on the other hand that do not have the executable bit set should have the extension `.sh` and no ``#!`` at the beginning of the script. This is an executable and therefore I would prefer it without file extension. But I can take over snake case if it makes you happy.

Comment (Araq) [reply tests/untestable/gdb/prettyPrinterTestInvoke]: I don't know why you think you need to lecture me about Unix's barbaric conventions. In Nim's repo filenames have an extension, just like in Nim code variables have a type.

**Comment by dom96** [comment]
Wouldn't this be better as a Nimble package?

**Comment by krux02** [comment]
@dom96 no, not at all. It is not a nim library. And the pretty printers are very tightly bound to the internal memory layout of the ``seq`` implementation. So when the implementation of ``seq`` changes the printers would need to be updated as well. I also don't like the idea that you would need to install a nimble package first, before you are able to debug Nim applications. No I think these pretty printers should be tested automatically by the Nim distribution and distributed by the Nim distribution.

File: tools/nim-gdb.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+  # workaround I just say I am not resposible for printing pointer
+  # types (seq and string are exception as they are semantically
+  # values).  this way the default type printer will handle pointer
+  # types and dive into the members of that type.  So I can still
+  # control with `ptype myval` and `ptype *myval` if I want to have
+  # detail or not.  I this this method stinks but I could not figure
+  # out a better solution.
+
+  type_map_static = {
+    'NI': 'int',  'NI8': 'int8', 'NI16': 'int16',  'NI32': 'int32',  'NI64': 'in64',
```
Comment (Araq) [inline tools/nim-gdb.py]: 'NI' needs to be mapped to 'long'

Comment (Araq) [inline tools/nim-gdb.py]: typo: 'in64'

File: tools/nim-gdb.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+    'NU': 'uint', 'NU8': 'uint8','NU16': 'uint16', 'NU32': 'uint32', 'NU64': 'uint64',
+    'NF': 'float', 'NF32': 'float32', 'NF32': 'float64',
```
Comment (Araq) [inline tools/nim-gdb.py]: 'NF32' twice?

Comment (krux02) [reply tools/nim-gdb.py]: that is a bug, thank you for pointing this out. Aparently I did not test of float64. But apart from that, why doesn't nim directly generate the fixed size types defined in `<stdint.h>`? `<nimbase.h>` could ensure that those types are always defined.

Comment (krux02) [reply tools/nim-gdb.py]: This is an error that the python environment could have told me.

**Comment by alehander42** [comment]
Can we force the compiler to emit the NTI types in debug mode? This way one can always have better info for enums etc

**Comment by krux02** [comment]
I would even go so far and emit debug information in debug mode. But currently one has to explicity pass ``--debugger:native`` to nim.

**Pull request merged**